File: go-sdk/pkg/observability/tracing/kafka_exporter.go
package tracing

import (
	"context"
	"os"
	"strings"
	"sync"
	"time"

	"github.com/IBM/sarama"
	"go.opentelemetry.io/otel/attribute"
	"go.opentelemetry.io/otel/sdk/trace"
	oteltrace "go.opentelemetry.io/otel/trace"
	"go.opentelemetry.io/proto/otlp/collector/trace/v1"
	commonpb "go.opentelemetry.io/proto/otlp/common/v1"
	resourcepb "go.opentelemetry.io/proto/otlp/resource/v1"
	tracepb "go.opentelemetry.io/proto/otlp/trace/v1"
	"google.golang.org/protobuf/proto"
)

// KafkaExporter exports OpenTelemetry spans to Kafka
type KafkaExporter struct {
	producer    sarama.AsyncProducer
	topic       string
	serviceName string
	mu          sync.Mutex
	closed      bool
}

// ExporterConfig holds exporter configuration
type ExporterConfig struct {
	ServiceName  string
	KafkaBrokers []string
	Topic        string
	EnableKafka  bool
}

// NewExporterConfig creates config from environment
func NewExporterConfig(serviceName string) *ExporterConfig {
	cfg := &ExporterConfig{
		ServiceName: serviceName,
		Topic:       getEnv("KAFKA_TRACES_TOPIC", "traces.application"),
		EnableKafka: getEnvBool("TRACES_KAFKA_ENABLED", true),
	}

	brokers := getEnv("KAFKA_BROKERS", "")
	if brokers != "" {
		cfg.KafkaBrokers = strings.Split(brokers, ",")
	}

	return cfg
}

// NewKafkaExporter creates a new Kafka span exporter
func NewKafkaExporter(cfg *ExporterConfig) (*KafkaExporter, error) {
	if len(cfg.KafkaBrokers) == 0 {
		return nil, nil
	}

	saramaConfig := sarama.NewConfig()
	saramaConfig.Producer.RequiredAcks = sarama.WaitForLocal
	saramaConfig.Producer.Compression = sarama.CompressionSnappy
	saramaConfig.Producer.Flush.Frequency = 100 * time.Millisecond
	saramaConfig.Producer.Flush.Messages = 100
	saramaConfig.Producer.Return.Errors = true

	producer, err := sarama.NewAsyncProducer(cfg.KafkaBrokers, saramaConfig)
	if err != nil {
		return nil, err
	}

	exp := &KafkaExporter{
		producer:    producer,
		topic:       cfg.Topic,
		serviceName: cfg.ServiceName,
	}

	go exp.handleErrors()

	return exp, nil
}

// ExportSpans exports spans to Kafka in OTLP protobuf format
func (e *KafkaExporter) ExportSpans(ctx context.Context, spans []trace.ReadOnlySpan) error {
	e.mu.Lock()
	if e.closed {
		e.mu.Unlock()
		return nil
	}
	e.mu.Unlock()

	// Convert to OTLP format
	otlpSpans := e.convertToOTLP(spans)
	
	// Marshal to protobuf
	data, err := proto.Marshal(otlpSpans)
	if err != nil {
		os.Stderr.WriteString("Failed to marshal OTLP traces: " + err.Error() + "\n")
		return err
	}

	// Use first span's trace ID as key
	var key string
	if len(spans) > 0 {
		key = spans[0].SpanContext().TraceID().String()
	}

	msg := &sarama.ProducerMessage{
		Topic:     e.topic,
		Key:       sarama.StringEncoder(key),
		Value:     sarama.ByteEncoder(data),
		Timestamp: time.Now(),
	}

	select {
	case e.producer.Input() <- msg:
	default:
		// Channel full, drop span
	}

	return nil
}

// Shutdown gracefully shuts down the exporter
func (e *KafkaExporter) Shutdown(ctx context.Context) error {
	e.mu.Lock()
	defer e.mu.Unlock()

	if e.closed {
		return nil
	}
	e.closed = true

	return e.producer.Close()
}

// convertToOTLP converts OpenTelemetry spans to OTLP format
func (e *KafkaExporter) convertToOTLP(spans []trace.ReadOnlySpan) *v1.ExportTraceServiceRequest {
	var otlpSpans []*tracepb.Span

	for _, span := range spans {
		sc := span.SpanContext()
		traceID := sc.TraceID()
		spanID := sc.SpanID()
		
		var parentSpanID []byte
		if span.Parent().IsValid() {
			pid := span.Parent().SpanID()
			parentSpanID = pid[:]
		}
		
		// Convert attributes
		var attributes []*commonpb.KeyValue
		for _, attr := range span.Attributes() {
			attributes = append(attributes, &commonpb.KeyValue{
				Key:   string(attr.Key),
				Value: attributeValueToOTLP(attr.Value),
			})
		}

		// Convert events
		var events []*tracepb.Span_Event
		for _, event := range span.Events() {
			var eventAttrs []*commonpb.KeyValue
			for _, attr := range event.Attributes {
				eventAttrs = append(eventAttrs, &commonpb.KeyValue{
					Key:   string(attr.Key),
					Value: attributeValueToOTLP(attr.Value),
				})
			}
			events = append(events, &tracepb.Span_Event{
				TimeUnixNano: uint64(event.Time.UnixNano()),
				Name:         event.Name,
				Attributes:   eventAttrs,
			})
		}

		// Convert status
		status := &tracepb.Status{
			Code: tracepb.Status_StatusCode(span.Status().Code),
			Message: span.Status().Description,
		}

		otlpSpan := &tracepb.Span{
			TraceId:           traceID[:],
			SpanId:            spanID[:],
			ParentSpanId:      parentSpanID,
			Name:              span.Name(),
			Kind:              tracepb.Span_SpanKind(span.SpanKind()),
			StartTimeUnixNano: uint64(span.StartTime().UnixNano()),
			EndTimeUnixNano:   uint64(span.EndTime().UnixNano()),
			Attributes:        attributes,
			Events:            events,
			Status:            status,
		}

		otlpSpans = append(otlpSpans, otlpSpan)
	}

	return &v1.ExportTraceServiceRequest{
		ResourceSpans: []*tracepb.ResourceSpans{
			{
				Resource: &resourcepb.Resource{
					Attributes: []*commonpb.KeyValue{
						{
							Key: "service.name",
							Value: &commonpb.AnyValue{
								Value: &commonpb.AnyValue_StringValue{
									StringValue: e.serviceName,
								},
							},
						},
					},
				},
				ScopeSpans: []*tracepb.ScopeSpans{
					{
						Spans: otlpSpans,
					},
				},
			},
		},
	}
}

// attributeValueToOTLP converts an attribute value to OTLP format
func attributeValueToOTLP(value attribute.Value) *commonpb.AnyValue {
	switch value.Type() {
	case attribute.BOOL:
		return &commonpb.AnyValue{
			Value: &commonpb.AnyValue_BoolValue{BoolValue: value.AsBool()},
		}
	case attribute.INT64:
		return &commonpb.AnyValue{
			Value: &commonpb.AnyValue_IntValue{IntValue: value.AsInt64()},
		}
	case attribute.FLOAT64:
		return &commonpb.AnyValue{
			Value: &commonpb.AnyValue_DoubleValue{DoubleValue: value.AsFloat64()},
		}
	case attribute.STRING:
		return &commonpb.AnyValue{
			Value: &commonpb.AnyValue_StringValue{StringValue: value.AsString()},
		}
	default:
		return &commonpb.AnyValue{
			Value: &commonpb.AnyValue_StringValue{StringValue: value.AsString()},
		}
	}
}

func (e *KafkaExporter) handleErrors() {
	for err := range e.producer.Errors() {
		os.Stderr.WriteString("Kafka trace exporter error: " + err.Error() + "\n")
	}
}

func getEnv(key, defaultValue string) string {
	if value := os.Getenv(key); value != "" {
		return value
	}
	return defaultValue
}

func getEnvBool(key string, defaultValue bool) bool {
	value := os.Getenv(key)
	if value == "" {
		return defaultValue
	}
	return value == "true" || value == "1" || value == "yes"
}

// GetTraceID extracts trace ID from context
func GetTraceID(ctx context.Context) string {
	spanCtx := oteltrace.SpanContextFromContext(ctx)
	if spanCtx.HasTraceID() {
		return spanCtx.TraceID().String()
	}
	return ""
}

File: python-sdk/observability/tracing/kafka_exporter.py
"""Kafka span exporter for OpenTelemetry."""

import os
import sys
import threading
from queue import Queue, Full
from typing import Optional, Sequence

from kafka import KafkaProducer
from kafka.errors import KafkaError
from opentelemetry.sdk.trace import ReadableSpan
from opentelemetry.sdk.trace.export import SpanExporter, SpanExportResult
from opentelemetry.proto.collector.trace.v1.trace_service_pb2 import (
    ExportTraceServiceRequest,
)
from opentelemetry.proto.common.v1.common_pb2 import AnyValue, KeyValue
from opentelemetry.proto.trace.v1.trace_pb2 import (
    ResourceSpans,
    ScopeSpans,
    Span,
    Status,
)
from opentelemetry.proto.resource.v1.resource_pb2 import Resource


class KafkaSpanExporter(SpanExporter):
    """Exports OpenTelemetry spans to Kafka in OTLP protobuf format."""
    
    def __init__(
        self,
        service_name: str,
        kafka_brokers: list[str],
        topic: str,
    ):
        self.service_name = service_name
        self.topic = topic
        self._producer: Optional[KafkaProducer] = None
        self._queue: Queue = Queue(maxsize=1000)
        self._closed = False
        self._lock = threading.Lock()
        
        self._init_producer(kafka_brokers)
        if self._producer:
            self._start_sender_thread()
    
    def _init_producer(self, kafka_brokers: list[str]):
        """Initialize Kafka producer."""
        try:
            self._producer = KafkaProducer(
                bootstrap_servers=kafka_brokers,
                acks=1,
                compression_type="gzip",
            )
        except KafkaError as e:
            sys.stderr.write(f"Failed to initialize trace Kafka producer: {e}\n")
            self._producer = None
    
    def _start_sender_thread(self):
        """Start background sender thread."""
        thread = threading.Thread(target=self._sender_loop, daemon=True)
        thread.start()
    
    def _sender_loop(self):
        """Background loop to send spans."""
        while not self._closed:
            try:
                span_data = self._queue.get(timeout=1.0)
                if span_data and self._producer:
                    self._producer.send(
                        self.topic,
                        key=span_data["key"],
                        value=span_data["value"],
                    )
            except Exception:
                pass
    
    def export(self, spans: Sequence[ReadableSpan]) -> SpanExportResult:
        """Export spans to Kafka in OTLP protobuf format."""
        if self._closed or not self._producer:
            return SpanExportResult.FAILURE
        
        # Convert to OTLP and serialize
        otlp_request = self._convert_to_otlp(spans)
        protobuf_data = otlp_request.SerializeToString()
        
        # Use first span's trace ID as key
        key = b""
        if spans:
            ctx = spans[0].get_span_context()
            key = format(ctx.trace_id, "032x").encode()
        
        try:
            self._queue.put_nowait({
                "key": key,
                "value": protobuf_data,
            })
        except Full:
            pass  # Drop if queue full
        
        return SpanExportResult.SUCCESS
    
    def shutdown(self) -> None:
        """Shutdown the exporter."""
        with self._lock:
            self._closed = True
            if self._producer:
                self._producer.flush()
                self._producer.close()
    
    def force_flush(self, timeout_millis: int = 30000) -> bool:
        """Force flush any pending spans."""
        if self._producer:
            self._producer.flush(timeout=timeout_millis / 1000)
        return True
    
    def _convert_to_otlp(self, spans: Sequence[ReadableSpan]) -> ExportTraceServiceRequest:
        """Convert spans to OTLP format."""
        otlp_spans = []
        
        for span in spans:
            ctx = span.get_span_context()
            parent = span.parent
            
            # Convert attributes
            attributes = []
            if span.attributes:
                for key, value in span.attributes.items():
                    attributes.append(
                        KeyValue(
                            key=key,
                            value=self._value_to_otlp(value)
                        )
                    )
            
            # Convert events
            events = []
            if span.events:
                for event in span.events:
                    event_attrs = []
                    if event.attributes:
                        for key, value in event.attributes.items():
                            event_attrs.append(
                                KeyValue(
                                    key=key,
                                    value=self._value_to_otlp(value)
                                )
                            )
                    events.append(
                        Span.Event(
                            time_unix_nano=event.timestamp,
                            name=event.name,
                            attributes=event_attrs,
                        )
                    )
            
            # Convert status
            status = Status(
                code=span.status.status_code.value,
                message=span.status.description or "",
            )
            
            # Create OTLP span
            otlp_span = Span(
                trace_id=ctx.trace_id.to_bytes(16, "big"),
                span_id=ctx.span_id.to_bytes(8, "big"),
                parent_span_id=parent.span_id.to_bytes(8, "big") if parent else b"",
                name=span.name,
                kind=span.kind.value + 1,  # OTLP enum starts at 1
                start_time_unix_nano=span.start_time,
                end_time_unix_nano=span.end_time or 0,
                attributes=attributes,
                events=events,
                status=status,
            )
            
            otlp_spans.append(otlp_span)
        
        # Create OTLP request
        return ExportTraceServiceRequest(
            resource_spans=[
                ResourceSpans(
                    resource=Resource(
                        attributes=[
                            KeyValue(
                                key="service.name",
                                value=AnyValue(string_value=self.service_name)
                            )
                        ]
                    ),
                    scope_spans=[
                        ScopeSpans(spans=otlp_spans)
                    ],
                )
            ]
        )
    
    def _value_to_otlp(self, value) -> AnyValue:
        """Convert a value to OTLP AnyValue."""
        if isinstance(value, bool):
            return AnyValue(bool_value=value)
        elif isinstance(value, int):
            return AnyValue(int_value=value)
        elif isinstance(value, float):
            return AnyValue(double_value=value)
        elif isinstance(value, str):
            return AnyValue(string_value=value)
        else:
            return AnyValue(string_value=str(value))
